runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -91,99 +70,47 @@
-        }
-
-        @Override
-        public void onDecompressionCompleted() {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.initXWalkLibrary();
-        }
-
-        @Override
-        public void onDownloadStarted() {
-            mXWalkActivity.showDialog(mXWalkActivity.getDownloadProgressDialog());
-        }
-
-        @Override
-        public void onDownloadUpdated(int percentage) {
-            ProgressDialog dialog = (ProgressDialog) mXWalkActivity.mActiveDialog;
-            dialog.setIndeterminate(false);
-            dialog.setMax(100);
-            dialog.setProgress(percentage);
-        }
 
-        @Override
-        public void onDownloadCancelled() {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.finish();
-        }
-
-        @Override
-        public void onDownloadCompleted(Uri uri) {
-            mXWalkActivity.dismissDialog();
-
-            Log.d(TAG, "Install the Crosswalk library, " + uri.toString());
-            Intent install = new Intent(Intent.ACTION_VIEW);
-            install.setDataAndType(uri, "application/vnd.android.package-archive");
-            mXWalkActivity.startActivity(install);
-        }
-
-        @Override
-        public void onDownloadFailed(int status, int error) {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.showDialog(mXWalkActivity.getDownloadFailedDialog(status, error));
-        }
-
-        @Override
-        public void onInitializationStarted() {
-        }
+    /**
+     * Runs on the UI thread to notify the Crosswalk runtime is ready
+     */
+    protected abstract void onXWalkReady();
 
-        @Override
-        public void onInitializationCompleted() {
-            mXWalkActivity.mIsXWalkReady = true;
-            mXWalkActivity.onXWalkReady();
-        }
+    /**
+     * True if the Crosswalk runtime is ready, false otherwise
+     */
+    protected boolean isXWalkReady() {
+        return mIsXWalkReady;
     }
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
-        mLibraryListener = new XWalkLibraryListener(this);
-        XWalkLibraryLoader.prepareToInit();
-        XWalkLibraryLoader.startDecompression(mLibraryListener, this);
+        Runnable completeCommand = new Runnable() {
+            @Override
+            public void run() {
+                mIsXWalkReady = true;
+                onXWalkReady();
+            }
+        };
+        Runnable cancelCommand = new Runnable() {
+            @Override
+            public void run() {
+                finish();
+            }
+        };
+
+        mActivityDelegate = new XWalkActivityDelegate(this, completeCommand, cancelCommand);
     }
 
     @Override
     protected void onResume() {
         super.onResume();
-        if (!mIsXWalkReady && !(mActiveDialog instanceof ProgressDialog)) initXWalkLibrary();
+
+        if (!mIsXWalkReady) mActivityDelegate.initialize();
```

Comment:
How about move mActivityDelegate.initialize() into onCreate?
Since we only need to invoke this once.
BTW, I think mIsXWalkReady could move into ActivityDelegate, since it's more related.